### PR TITLE
Config prefer upstream

### PR DESCRIPTION
**Pull request by Paul-Ferrell**
This allows for quickly configuring a spack install/env to use upstream packages by default. This is particularly important when upstreaming from a set of officially supported spack installs on a production cluster. By configuring such that package preferences match the upstream, you ensure maximal reuse of existing package installations.

### CODE REVIEW

**Comment by Paul-Ferrell** [comment]
Ok that last update should take care of code coverage. The only error left will be clingo, and that failure seems unrelated to this merge request.

**Comment by alalazo** [comment]
@Paul-Ferrell Can you write a bit about the use case that originated this PR?

**Comment by Paul-Ferrell** [comment]
This allows for quickly configuring a spack install/env to use upstream packages by default. This is particularly important when upstreaming from a set of officially supported spack installs on a production cluster. By configuring such that package preferences match the upstream, you ensure maximal reuse of existing package installations.

File: lib/spack/spack/cmd/config.py [hunk 1]
```diff
@@ -431,6 +444,92 @@ def config_revert(args):
...
 
+def config_prefer_upstream(args):
+    """Generate a packages config based on the configuration of all upstream
+    installs."""
+
+    scope = args.scope
+    if scope is None:
+        scope = spack.config.default_modify_scope('packages')
+
+    specs = spack.store.db.query(installed=[InstallStatuses.INSTALLED])
```
Comment (becker33) [inline lib/spack/spack/cmd/config.py]: This will avoid the need to import from the database
```suggestion
    specs = spack.store.db.query(installed=True)
```

File: lib/spack/spack/cmd/config.py [hunk 2]
```diff
@@ -431,6 +444,92 @@ def config_revert(args):
...
+            upstream = spec.package.installed_upstream
+        except spack.repo.UnknownNamespaceError as err:
+            tty.die(
+                "Could not find package when checking spec {0} ({1}). "
+                "This is usually due to your Spack instance not being "
+                "configured to know about the upstream's repositories."
+                .format(spec.name, err.message)
+            )
+        if (upstream and not args.local) or (not upstream and args.local):
+            pref_specs.append(spec)
```
Comment (becker33) [inline lib/spack/spack/cmd/config.py]: You can avoid the corner case of a missing repo by diffing `spack.store.db.query(...)` and `spack.store.db.query_local(...)` results to get the list of specs to prefer. That way you don't have to query the package for each one, and the code is much smaller. I think it also leads to fewer DB reads since we do a DB operation for every check of `Package.installed_upstream`.

```suggestion
    all_specs = set(spack.store.db.query(installed=True))
    local_specs = set(spack.store.db.query_local(installed=True))
    pref_specs = local_specs if args.local else all_specs - local_spec
```

File: lib/spack/spack/test/cmd/config.py [hunk 3]
```diff
@@ -645,3 +648,49 @@ def check_config_updated(data):
     assert isinstance(data['install_tree'], dict)
     assert data['install_tree']['root'] == '/fake/path'
     assert data['install_tree']['projections'] == {'all': '{name}-{version}'}
+
+
+def test_config_prefer_upstream(tmpdir_factory, install_mockery, mock_fetch,
```
Comment (becker33) [inline lib/spack/spack/test/cmd/config.py:653]: This test needs the `mutable_config` fixture so that failures won't leave detritus in real config directories

File: lib/spack/spack/test/cmd/config.py [hunk 4]
```diff
@@ -645,3 +648,49 @@ def check_config_updated(data):
...
+    assert packages['boost'] == {
+        'compiler': ['gcc@4.5.0'],
+        'variants': '+debug +graph',
+        'version': ['1.63.0']}
+    assert packages['dependency-install'] == {
+        'compiler': ['gcc@4.5.0'], 'version': ['2.0']}
+    assert packages['hdf5'] == {
+        'compiler': ['gcc@4.5.0'], 'version': ['2.3']}
+
+    assert '- hdf5' in output
```
Comment (becker33) [inline lib/spack/spack/test/cmd/config.py:697]: Can you add a comment here explaining what it's testing? I get that it's testing that hdf5 is correctly identified as having conflicting variants, but it's hard to parse for a future maintainer.

File: lib/spack/spack/test/cmd/config.py [hunk 5]
```diff
@@ -645,3 +648,49 @@ def check_config_updated(data):
...
+    db_for_test = spack.database.Database(
+        downstream_db_root, upstream_dbs=[prepared_db])
+    monkeypatch.setattr(spack.store, 'db', db_for_test)
+
+    output = config('prefer-upstream')
+    scope = spack.config.default_modify_scope('packages')
+    cfg_file = spack.config.config.get_config_filename(scope, 'packages')
+    packages = syaml.load(open(cfg_file))['packages']
+    # Cleanup the config file we added, so it doesn't break other unit tests.
+    os.unlink(cfg_file)
```
Comment (becker33) [inline lib/spack/spack/test/cmd/config.py]: This can go away once we use `mutable_config` fixture

**Review by becker33 — CHANGES_REQUESTED** [review]
I keep thinking there has to be a better way to handle the variant conflicts, but I can't think of one when I actually get down to it so this might be best. A couple requests for the test and how you construct the spec list.

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/cmd/config.py [hunk 6]
```diff
@@ -431,6 +443,79 @@ def config_revert(args):
...
+    """Generate a packages config based on the configuration of all upstream
+    installs."""
+
+    scope = args.scope
+    if scope is None:
+        scope = spack.config.default_modify_scope('packages')
+
+    all_specs = set(spack.store.db.query(installed=True))
+    local_specs = set(spack.store.db.query_local(installed=True))
+    pref_specs = local_specs if args.local else all_specs - local_spec
```
Comment (becker33) [inline lib/spack/spack/cmd/config.py]:
```suggestion
    pref_specs = local_specs if args.local else all_specs - local_specs
```

**Review by becker33 — CHANGES_REQUESTED** [review]

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/cmd/config.py [hunk 7]
```diff
@@ -431,6 +443,79 @@ def config_revert(args):
...
+        # Collect all the upstream compilers and versions for this package.
+        pkg = pkgs.get(spec.name, {
+            'version': [],
+            'compiler': [],
+        })
+        pkgs[spec.name] = pkg
+
+        # We have no existing variant if this is our first added version.
+        existing_variants = pkg.get('variants',
+                                    None if not pkg['version'] else '')
```
Comment (becker33) [inline lib/spack/spack/cmd/config.py:471]:
```suggestion
                                    None)
```
Below you check whether it's None, not whether it's False-ish

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/config.py (hunk 7 above)
Comment (Paul-Ferrell) [reply lib/spack/spack/cmd/config.py:471]: This is intentional. None denotes that this is the first time we've seen this package, while an empty string is simply the lack of variants if there weren't any.

Comment (becker33) [reply lib/spack/spack/cmd/config.py:471]: Intentionally conflicting no variants specified with any specified variants?

hmm... so we tried none-specified as if they specifically requested all the defaults... I guess that works well for this use case, just not what we usually do.

Comment (becker33) [reply lib/spack/spack/cmd/config.py:471]: Ok, then I see the other way to get rid of the test failure below. I'll suggest it inline.

File: lib/spack/spack/cmd/config.py [hunk 8]
```diff
@@ -431,6 +443,79 @@ def config_revert(args):
...
+                variants.append(str(variant))
+        variants.sort()
+        variants = ' '.join(variants)
+
+        if spec.name not in conflicting_variants:
+            # Only specify the variants if there's a single variant
+            # set across all versions/compilers.
+            if existing_variants is not None and existing_variants != variants:
+                conflicting_variants.add(spec.name)
+                del pkg['variants']
```
Comment (becker33) [inline lib/spack/spack/cmd/config.py]:
```suggestion
                pkg.pop('variants')
```
If the first variants setting was empty, we don't have anything to delete from the dictionary when we reach a conflict with the second one.

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/config.py [hunk 9]
```diff
@@ -431,6 +443,79 @@ def config_revert(args):
...
+                variants.append(str(variant))
+        variants.sort()
+        variants = ' '.join(variants)
+
+        if spec.name not in conflicting_variants:
+            # Only specify the variants if there's a single variant
+            # set across all versions/compilers.
+            if existing_variants is not None and existing_variants != variants:
+                conflicting_variants.add(spec.name)
+                pkg.pop('variants')
```
Comment (becker33) [inline lib/spack/spack/cmd/config.py]:
```suggestion
                pkg.pop('variants', None)
```

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/cmd/config.py (hunk 7 above)
Comment (Paul-Ferrell) [reply lib/spack/spack/cmd/config.py:471]: Yes, the variants string would be an empty string if no variants differed from the defaults. 
The goal is to make a concise configuration, not an exhaustive one.
